src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,951 @@
+                ? generator.GenericName(methodSymbol.Name, methodSymbol.TypeArguments)
+                : generator.IdentifierName(methodSymbol.Name);
+            var invocation = generator.InvocationExpression(memberName, arguments);
+
+            SyntaxNode newStatement;
+            if (!methodSymbol.ReturnsVoid)
+            {
+                newStatement = generator.ReturnStatement(invocation);
+            }
+            else
+            {
+                newStatement = generator.ExpressionStatement(invocation);
+            }
+
+            var codeGenerationService = document.GetRequiredLanguageService<ICodeGenerationService>();
+            var options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var newMethod = CodeGenerationSymbolFactory.CreateMethodSymbol(methodSymbol, statements: SpecializedCollections.SingletonEnumerable(newStatement).ToImmutableArray(), containingType: methodSymbol.ContainingType);
+            var newMethodDeclaration = codeGenerationService.CreateMethodDeclaration(newMethod, options: new CodeGenerationOptions(options: options, parseOptions: expression.SyntaxTree.Options));
+            return newMethodDeclaration;
+        }
+
+        /// <summary>
+        /// This method goes through all the invocation sites and adds a new argument with the expression to be added.
+        /// It also introduces a parameter at the original method site.
+        /// 
+        /// Example:
+        /// public void M(int x, int y)
+        /// {
+        ///     int f = [|x * y|];
+        /// }
+        /// 
+        /// public void InvokeMethod()
+        /// {
+        ///     M(5, 6);
+        /// }
+        /// 
+        /// ---------------------------------------------------->
+        /// 
+        /// public void M(int x, int y, int f) // parameter gets introduced
+        /// {
+        /// }
+        /// 
+        /// public void InvokeMethod()
+        /// {
+        ///     M(5, 6, 5 * 6); // argument gets added to callsite
+        /// }
+        /// </summary>
+        private async Task<SyntaxNode> ModifyDocumentInvocationsAndIntroduceParameterAsync(Compilation compilation,
+            Document originalDocument, Document document, Dictionary<TIdentifierNameSyntax, IParameterSymbol> mappingDictionary,
+            IMethodSymbol methodSymbol, SyntaxNode containingMethod, TExpressionSyntax expression,
+            bool allOccurrences, string parameterName,
+            List<SyntaxNode> invocations, CancellationToken cancellationToken)
+        {
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var semanticFacts = document.GetRequiredLanguageService<ISemanticFactsService>();
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var editor = new SyntaxEditor(root, generator);
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
```

Comment:
it's a nit, but i would really like if you kept associated concepts together.  so, for example, acquire SyntaxFacts and SemanticFacts at the same time.  Get teh Compilation/SemanticMode/Root at the same time.  Get the generator and editor at the same time.  etc.

Reply:
makes sense